design/baremetal-operator/firmware-image-update.md

```diff
@@ -0,0 +1,195 @@
+# Firmware Image Update
+
+## Status
+
+provisional
+
+## Summary
+
+Provide the ability to get and update vendor firmware version.
+
+## Motivation
+
+It desired to update vendor firmware before deploying the OS image of the host.
+
+### Goals
+
+- Agree on the format of the vendor neutral API and attributes
+- Implemented Nvidia firmware update for NICs
+
+### Non-Goals
+
+- Implemented OOB firmware update for ILO and redfish
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Cloud admin wants to update Nvidia NICS firmware before OS deployed on bare-metal host.
+
+## Design Details
+
+### New API (CRD) Proposal
+
+We propose to introduce a new Custom Resource Definition (CRD) `HostFirmwareImage`. `HostFirmwareImage` will store a list of firmware images to be applied (if required) on the bare-metal host. The `HostFirmwareImage` will be mapped to Baremetal Host object by namespace and name. The spec section will introduce the following attributes:
+
+- url - location to download the firmware image.
+- checksum - checksum of the firmware image. Optional attribute.
+- checksumType - checksum algorithm for the firmware image can be md5, sha256, sha512. Optional attribute.
+- vendor – The hardware vendor which is responsible for that firmware. Current proposal it will be only nvidia and can be extended in future proposals for other vendors.
+- component – The hardware component which this firmware matches. Hardware vendor may have several types of hardware for example Nvidia has NiC’s and GPU. Current proposal will be only `net` and can be extended in future proposals for other vendors.
+- componentFlavor- indicate the component flavor. Optional attribute. For Nvidia NICs it will be the PSID.
+- version - the requested firmware version of the firmware image.
+
+example of the HostFirmwareImage spec section:
+
+```yaml
+spec:
+  firmwareImages:
+    - url: http://firmware_images/fw1.bin
+      checksum: a94e683ea16d9ae44768f0a65942234d
+      checksumType: md5
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_1090120019
+      version: 24.34.1002

```

Comment:
It's unfortunate that the user needs to correctly describe the version that corresponds to the URL, but I can't think of a way around that.